tests/unit/s2n_map_iterator_test.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/s2n.h"
+#include "s2n_test.h"
+#include "utils/s2n_map.h"
+#include "utils/s2n_map_internal.h"
+
+#define TEST_VALUE_COUNT 10
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* s2n_map_iterator iteration test */
+    {
+        struct s2n_map *map = s2n_map_new();
+        EXPECT_NOT_NULL(map);
+        /* fail to initialize an iterator on a mutable map */
+        {
+            struct s2n_map_iterator iter = { 0 };
+            EXPECT_ERROR_WITH_ERRNO(s2n_map_iterator_init(&iter, map), S2N_ERR_MAP_MUTABLE);
+        };
+
+        EXPECT_OK(s2n_map_complete(map));
+
+        /* next returns an error when the blob is null */
+        {
+            struct s2n_map_iterator iter = { 0 };
+            EXPECT_OK(s2n_map_iterator_init(&iter, map));
+
+            EXPECT_ERROR_WITH_ERRNO(s2n_map_iterator_next(&iter, NULL), S2N_ERR_SAFETY);
```

Comment:
I think you've got a minor bug here: the error should be S2N_ERR_NULL. I think you're instead getting S2N_ERR_SAFETY because the map is empty so there is no next and you don't get far enough in the function for the NULL to matter. This is why error types are important.

Maybe use S2N_ERR_ARRAY_INDEX_OOB for the case where there is no next element? Or add a new internal error type. That'll make your errors clearer. S2N_ERR_SAFETY can be tricky.

If you don't want to use more specific errors, the other method to make sure you're actually hitting the right failure is to have a control case. So, call s2n_map_iterator_next without a NULL and check that it succeeds. Then you can be more confident that the NULL was the problem.